src/infra/dex/balancer/dto.rs

```diff
@@ -1,58 +1,208 @@
+        };
+        Ok(Self {
+            query: get_swap_paths_query::QUERY,
+            variables,
+        })
+    }
 }
 
-/// An SOR query.
+/// Ether amount refers to the SOR API V3's `AmountHumanReadable` type.
 #[serde_as]
-#[derive(Debug, Serialize)]
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct EtherAmount(BigDecimal);
+
+impl EtherAmount {
+    pub fn from_wei(wei: &U256) -> EtherAmount {
+        Self(u256_to_big_decimal(wei) / BigDecimal::from(10_u64.pow(18)))
+    }
+
+    pub fn value(&self) -> BigDecimal {
+        self.0.clone()
+    }
+
+    pub fn to_wei(&self) -> Option<U256> {
+        big_decimal_to_u256(&(&self.0 * BigDecimal::from(10_u64.pow(18))))
+    }
+}
+
+#[serde_as]
+#[derive(Serialize)]
 #[serde(rename_all = "camelCase")]
-pub struct Query {
-    /// The sell token to quote.
-    pub sell_token: H160,
-    /// The buy token to quote.
-    pub buy_token: H160,
-    /// The order kind to use.
-    pub order_kind: OrderKind,
-    /// The amount to quote
-    ///
-    /// For sell orders this is the exact amount of sell token to trade, for buy
-    /// orders, this is the amount of buy tokens to buy.
-    #[serde_as(as = "serialize::U256")]
-    pub amount: U256,
-    /// The current gas price estimate used for determining how the trading
-    /// route should be split.
-    #[serde_as(as = "serialize::U256")]
-    pub gas_price: U256,
+struct Variables {
+    call_data_input: CalDataInput,
+    /// The Chain to query.
+    chain: Chain,
+    /// Whether to run `queryBatchSwap` to update the return amount with most
+    /// up-to-date on-chain values.
+    query_batch_swap: bool,
+    /// The amount to swap in human form.
+    swap_amount: EtherAmount,
+    /// SwapType either exact_in or exact_out (also givenIn or givenOut).
+    swap_type: SwapType,
+    /// Token address of the tokenIn.
+    token_in: H160,
+    /// Token address of the tokenOut.
+    token_out: H160,
+    /// Which vault version to use. If none provided, will chose the better
+    /// return from either version.
+    use_vault_version: u8,
 }
 
-impl Query {
-    pub fn from_domain(order: &dex::Order, gas_price: auction::GasPrice) -> Self {
-        Self {
-            sell_token: order.sell.0,
-            buy_token: order.buy.0,
-            order_kind: match order.side {
-                order::Side::Buy => OrderKind::Buy,
-                order::Side::Sell => OrderKind::Sell,
-            },
-            amount: order.amount().amount,
-            gas_price: gas_price.0 .0,
+/// Inputs for the call data to create the swap transaction. If this input is
+/// given, call data is added to the response.
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct CalDataInput {
+    /// How long the swap should be valid, provide a timestamp. `999999999` for
+    /// infinite. Default: infinite.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    deadline: Option<u64>,
```

Comment:
Used infinite by default. Not sure if it is possible to calculate a reasonable value here. Or maybe we should use a configured constant value.

Reply:
You can use the current timestamp + 2 minutes. Also I think all requests might have a deadline by now which we could pipe into here as well. But overall we should pick a time that is short but long enough to be usable during an auction.

Reply:
I hade to open a separate PR in order to support it https://github.com/gnosis/solvers/pull/39